sdk/metric/config.go

```diff
@@ -14,12 +14,120 @@
+// will have their force-flush and shutdown methods unified into returned
+// single functions.
+func (c config) readerSignals() (forceFlush, shutdown func(context.Context) error) {
+	var fFuncs, sFuncs []func(context.Context) error
+	for r := range c.readers {
+		sFuncs = append(sFuncs, r.Shutdown)
+		fFuncs = append(fFuncs, r.ForceFlush)
+	}
+
+	return unify(fFuncs), unifyShutdown(sFuncs)
+}
+
+// unify unifies calling all of funcs into a single function call. It will
+// return nil if funcs is empty. All errors returned from calls to funcs will
+// be unify into a single error return value.
+func unify(funcs []func(context.Context) error) func(context.Context) error {
+	if len(funcs) == 0 {
+		return func(context.Context) error { return nil }
+	}
+
+	return func(ctx context.Context) error {
+		var errs []error
+		for _, f := range funcs {
+			if err := f(ctx); err != nil {
+				errs = append(errs, err)
+			}
+		}
+		switch len(errs) {
+		case 0:
+			return nil
+		case 1:
+			return errs[0]
+		default:
+			return fmt.Errorf("%v", errs)
+		}
+	}
+}
+
+// unifyShutdown unifies calling all of funcs once for a shutdown. If called
+// more than once, an ErrReaderShutdown error is returned.
+func unifyShutdown(funcs []func(context.Context) error) func(context.Context) error {
+	f := unify(funcs)
+	var once sync.Once
+	return func(ctx context.Context) error {
+		err := ErrReaderShutdown
+		once.Do(func() { err = f(ctx) })
+		return err
+	}
+}
+
+// newConfig returns a config configured with options.
+func newConfig(options []Option) config {
+	conf := config{res: resource.Default()}
+	for _, o := range options {
+		conf = o.apply(conf)
+	}
+	return conf
+}
 
 // Option applies a configuration option value to a MeterProvider.
 type Option interface {
 	apply(config) config
 }
 
-// TODO (#2819): implement provider options.
+// optionFunc applies a set of options to a config.
+type optionFunc func(config) config
+
+// apply returns a config with option(s) applied.
+func (o optionFunc) apply(conf config) config {
+	return o(conf)
+}
+
+// WithResource associates a Resource with a MeterProvider. This Resource
+// represents the entity producing telemetry and is associated with all Meters
+// the MeterProvider will create.
+//
+// By default, if this Option is not used, the default Resource from the
+// go.opentelemetry.io/otel/sdk/resource package will be used.
+func WithResource(res *resource.Resource) Option {
+	return optionFunc(func(conf config) config {
+		conf.res = res
+		return conf
+	})
+}
+
+// WithReader associates a Reader with a MeterProvider. Any passed view
+// options will be used to associate a view with the Reader. If no options are
+// passed the default view will be use for the Reader.
+//
+// By default, if this option is not used, the MeterProvider will perform no
+// operations; no data will be exported without a Reader.
+func WithReader(r Reader, opts ...view.Option) Option {
```

Comment:
I was wondering the same when I wrote this, but didn't have a good example of when passing multiple options wouldn't suffice.

The only downside I see is this will require an extra line to create.

```go
c := view.New(/* view opts */)
r := NewManualReader()
mp := NewMeterProvider(WithReader(r, c))
```

Instead of 

```go
r := NewManualReader()
mp := NewMeterProvider(WithReader(r, /* view opts */))
```

Though this is a moot point if we cannot express the needed view configuration. Are the options to define a view that reaggeragates the instrument "something.http" to a histogram and one that matches the redis Meter and only keeps endpoint labels going to be incompatible? Will it error if they are passed together?

If so, I'm happy to change to `func WithReader(r Reader, confs ...view.Config) Option`.

Reply:
The way the current options are laid out you can specify different match attributes, which are and'ed together, and 1 set of transforms, like reaggergation or remove labels.  So we could describe a view that selects a specific instrument AND meter which should apply reaggeragtion AND filter labels, but I can't describe something separate streams.

So if
```
A = Select an instrument
B = Select a meter
X = Change aggergation
Y = Filter labels
```

The current WithReader  could describe (A AND B + X AND Y) => reader, but it can't do (A+X OR B+Y) => reader.

So how I would expect it to be used is
```go
WithReader(rdr,
    view.New(
        view.MatchInstrument(A),
        view.WithAggregation(X),
    ),
    view.New(
        view.MatchMeter(B),
        view.WithLabels(Y),
    ),
)
```

Reply:
If we wanted to, there could be an option that is something like `WithMutliView()`. That would mean we would have to store in readers an interface instead of a struct view.Config, and that interface assume that one input instrument can result in multiple outputs.

Reply:
I think your example helped clarify things for me. I'll update to the alternate signature.